url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/org/jbox2d/collision/Collision.java

```diff
@@ -0,0 +1,844 @@
+	 * @return
+	 */
+	public static final int clipSegmentToLine(final ClipVertex[] vOut, final ClipVertex[] vIn, final Vec2 normal,
+			float offset) {
+		
+		// Start with no output points
+		int numOut = 0;
+		
+		// Calculate the distance of end points to the line
+		float distance0 = Vec2.dot(normal, vIn[0].v) - offset;
+		float distance1 = Vec2.dot(normal, vIn[1].v) - offset;
+		
+		// If the points are behind the plane
+		if (distance0 <= 0.0f) {
+			vOut[numOut++].set(vIn[0]);
+		}
+		if (distance1 <= 0.0f) {
+			vOut[numOut++].set(vIn[1]);
+		}
+		
+		// If the points are on different sides of the plane
+		if (distance0 * distance1 < 0.0f) {
+			// Find intersection point of edge and plane
+			float interp = distance0 / (distance0 - distance1);
+			// vOut[numOut].v = vIn[0].v + interp * (vIn[1].v - vIn[0].v);
+			vOut[numOut].v.set(vIn[1].v).subLocal(vIn[0].v).mulLocal(interp).addLocal(vIn[0].v);
+			if (distance0 > 0.0f) {
+				vOut[numOut].id.set(vIn[0].id);
+			}
+			else {
+				vOut[numOut].id.set(vIn[1].id);
+			}
+			++numOut;
+		}
+		
+		return numOut;
+	}
+	
+	// #### COLLISION STUFF (not from collision.h or collision.cpp) ####
+	
+	// djm pooling
+	
+	/**
+	 * Compute the collision manifold between two circles.
+	 * 
+	 * @param manifold
+	 * @param circle1
+	 * @param xfA
+	 * @param circle2
+	 * @param xfB
+	 */
+	public final void collideCircles(Manifold manifold, final CircleShape circle1, final Transform xfA,
+			final CircleShape circle2, final Transform xfB) {
+		manifold.pointCount = 0;
+		
+    // before inline:
+//		Transform.mulToOut(xfA, circle1.m_p, pA);
+//    Transform.mulToOut(xfB, circle2.m_p, pB);
+//		d.set(pB).subLocal(pA);
+//    float distSqr = d.x * d.x + d.y * d.y;
+    
+    // after inline:
+    final Vec2 v = circle1.m_p;
+		final float pAy = xfA.position.y + xfA.R.col1.y * v.x + xfA.R.col2.y * v.y;
+    final float pAx = xfA.position.x + xfA.R.col1.x * v.x + xfA.R.col2.x * v.y;
+
+    final Vec2 v1 = circle2.m_p;
+		final float pBy = xfB.position.y + xfB.R.col1.y * v1.x + xfB.R.col2.y * v1.y;
+    final float pBx = xfB.position.x + xfB.R.col1.x * v1.x + xfB.R.col2.x * v1.y;
+    
+    final float dx = pBx - pAx;
+    final float dy = pBy - pAy;
+    
+    final float distSqr = dx * dx + dy * dy;
+    // end inline
+		
+		final float radius = circle1.m_radius + circle2.m_radius;
+		if (distSqr > radius * radius) {
+			return;
+		}
+		
+		manifold.type = ManifoldType.CIRCLES;
+		manifold.localPoint.set(circle1.m_p);
+		manifold.localNormal.setZero();
+		manifold.pointCount = 1;
+		
+		manifold.points[0].localPoint.set(circle2.m_p);
+		manifold.points[0].id.zero();
+	}
+	
+	// djm pooling, and from above
+	
+	/**
+	 * Compute the collision manifold between a polygon and a circle.
+	 * 
+	 * @param manifold
+	 * @param polygon
+	 * @param xfA
+	 * @param circle
+	 * @param xfB
+	 */
+	public final void collidePolygonAndCircle(Manifold manifold, final PolygonShape polygon, final Transform xfA,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 89 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 304-324, lines 335-421 into separate methods.